cookbooks/cdo-apps/libraries/cdo_apps.rb

```diff
@@ -34,49 +33,71 @@ def setup_app(app_name)
-      mode '0755'
-
-      variables app_name: app_name
-      source "#{app_server}.service.erb"
+    directory socket_path do
+      action :create
+      owner user
+      recursive true
     end
 
-    template init_script do
+    # Set up a SystemD service to monitor the web server process.
+    template unit_file do
       app_server = node['cdo-apps']['app_server']
       src_file = "#{app_root}/config/#{app_server}.rb"
+      export_env = node['cdo-apps']['bundle_env'].merge(node['cdo-apps'][app_name]['env'] || {})
 
-      source "#{app_server}.sh.erb"
       user 'root'
       group 'root'
       mode '0755'
-      variables src_file: src_file,
-        app_name: app_name,
+
+      variables app_name: app_name,
         app_root: app_root,
-        pid_file: "#{src_file}.pid",
-        socket_path: socket_path,
-        user: user,
         env: node.chef_environment,
-        export_env: node['cdo-apps']['bundle_env'].
-          merge(node['cdo-apps'][app_name]['env'] || {})
-      notifies :reload, "service[#{app_name}]", :delayed
+        export_env: export_env,
+        src_file: src_file,
+        user: user
+      source "#{app_server}.service.erb"
+
+      notifies :run, "execute[restart #{app_name} service]", :delayed
     end
 
-    # Stop before and start after changes to significant init-script attributes.
-    file "#{app_name}_app_server" do
+    # Define an execute resource for restarting (or starting) the entire
+    # SystemD service, which can be invoked by other Chef resources
+    execute "restart #{app_name} service" do
+      command "systemctl restart #{app_name}"
+
+      # Don't run by default; rely on notifications.
       action :nothing
-      path "#{Chef::Config[:file_cache_path]}/#{app_name}_app_server"
-      content "#{node['cdo-apps']['app_server']}:#{socket_path}"
 
-      # Stop service before init-script update, and start service afterwards.
-      subscribes :create, "template[#{init_script}]", :before
-      notifies :stop, "service[#{app_name}]", :immediately
-      notifies :start, "service[#{app_name}]", :delayed
+      # Restart when Ruby is upgraded.
+      subscribes :run, "apt_package[ruby#{node['cdo-ruby']['version']}]", :delayed if node['cdo-ruby']
+
+      # Reload when gem bundle is updated.
+      subscribes :run, 'execute[bundle-install]', :delayed
+
+      # Reload when application is rebuilt.
+      subscribes :run, 'execute[build-cdo]', :delayed
+
+      # Reload when global config is updated to pick up changes.
+      subscribes :run, 'template[globals]', :delayed
+
+      # Ensure globals.yml is up-to-date before (re)starting service.
+      notifies :create, 'template[globals]', :before
+
+      only_if {File.exist? unit_file}
+    end
+
+    # We always want to restart the web server process whenever port/socket
+    # listener configuration is changed, so create a file with that information
+    # as its contents and invoke the restart whenever those contents change.
```

Comment:
We specify listener via code and config that in practice only changes with a release and an invocation of our build process (which stops/starts the service). Do we need to support this use case?

Reply:
It might not be strictly necessary, but building in protections for execution scenarios other than a full `ci_build` is still good practice. We don't want our Chef scripts to break unless they're run in a particular way by a particular external build process; they should be internally consistent.